src/ray/common/ray_syncer/ray_syncer-inl.h

```diff
@@ -136,58 +143,123 @@ class NodeSyncConnection {
   /// We'll filter the received or sent messages when the message is stale.
   absl::flat_hash_map<std::string, std::array<int64_t, kComponentArraySize>>
       node_versions_;
+
+  bool sending_ = false;
 };
 
-/// SyncConnection for gRPC server side. It has customized logic for sending.
-class ServerSyncConnection : public NodeSyncConnection {
+template <typename T>
+class BidiReactor : public T, public NodeSyncConnection {
  public:
-  ServerSyncConnection(
+  BidiReactor(
       instrumented_io_context &io_context,
       const std::string &remote_node_id,
-      std::function<void(std::shared_ptr<RaySyncMessage>)> message_processor);
+      std::function<void(std::shared_ptr<const RaySyncMessage>)> message_processor,
+      std::function<void(const std::string &, bool)> cleanup_cb)
+      : NodeSyncConnection(io_context,
+                           remote_node_id,
+                           std::move(message_processor),
+                           std::move(cleanup_cb)) {}
+
+  using T::StartRead;
+  using T::StartWrite;
+
+  void Send(std::shared_ptr<const RaySyncMessage> message, bool flush) override {
+    sending_message_ = std::move(message);
+    grpc::WriteOptions opts;
+    if (flush) {
+      opts.clear_buffer_hint();
+    } else {
+      opts.set_buffer_hint();
+    }
+    RAY_LOG(DEBUG) << "[BidiReactor] Sending message to "
+                   << NodeID::FromBinary(GetRemoteNodeID()) << " about node "
+                   << NodeID::FromBinary(sending_message_->node_id());
+    StartWrite(sending_message_.get(), opts);
+  }
 
-  ~ServerSyncConnection() override;
+  void StartPull() {
+    receiving_message_ = std::make_shared<RaySyncMessage>();
+    RAY_LOG(DEBUG) << "Start reading: " << NodeID::FromBinary(GetRemoteNodeID());
+    StartRead(receiving_message_.get());
+  }
 
-  void HandleLongPollingRequest(grpc::ServerUnaryReactor *reactor,
-                                RaySyncMessages *response);
+  void OnWriteDone(bool ok) override {
+    if (ok) {
+      io_context_.dispatch([this]() { SendNext(); }, "");
+    } else {
+      RAY_LOG(ERROR) << "Failed to send the message to: "
```

Comment:
should we use RAY_LOG_EVERY... in case the log is too noisy on failure?

Reply:
do we need to retry here?

Reply:
We shouldn't retry here, a failure means streaming failure and we need to reconstruct the data here which is done in OnDone.